example/Binary.hs

```diff
@@ -0,0 +1,389 @@
+subo a b c = addo b c a
+
+lesso :: Term Binary -> Term Binary -> Goal ()
+lesso a b = do
+  x <- fresh
+  poso x
+  addo a x b
+
+mulo :: Term Binary -> Term Binary -> Term Binary -> Goal ()
+mulo a b c =
+  disjMany
+    [ do
+        a === inject' []
+        c === a
+    , do
+        b === inject' []
+        c === b
+        poso a
+    , do
+        a === inject' [I]
+        b === c
+    , do
+        (ar, cr) <- fresh
+        a === Value (LogicCons (inject' O) ar)
+        c === Value (LogicCons (inject' O) cr)
+        poso b
+        poso ar
+        poso cr
+        mulo ar b cr
+    , do
+        ar <- fresh
+        a === Value (LogicCons (inject' I) ar)
+        poso b
+        poso ar
+        gtlo c
+
+        c1 <- fresh
+        lessl3o c1 c a b
+
+        mulo ar b c1
+        addo (Value (LogicCons (inject' O) c1)) b c
+    ]
+
+{- FOURMOLU_DISABLE -}
+divo :: Term Binary -> Term Binary -> Term Binary -> Term Binary -> Goal ()
+divo n m q r =
+  disjMany
+    [ do
+        q === inject' []
+        n === r
+        lesso n m
+    , do
+        q === inject' [I]
+        samelo n m
+        binaryo n
+        addo r m n
+        lesso r m
+    , do
+        lesslo m n
+        lesso r m
+        poso q
+
+        (n1, n2) <- fresh
+        splito n r n1 n2
+        (q1, q2) <- fresh
+        splito q r q1 q2
+        q2m <- fresh
+        n1 & (matche'
+          & on' _LogicNil' (\() -> do
+              q1 === inject' []
+              subo n2 r q2m
+              mulo q2 m q2m)
+          & on' _LogicCons' (\_ -> do
+              (q2mr, rr, r1) <- fresh
+              mulo q2 m q2m
+              addo q2m r q2mr
+              subo q2mr n2 rr
+              splito rr r r1 (inject' [])
+              divo n1 m q1 r1)
+          & enter')
+    ]
+{- FOURMOLU_ENABLE -}
+
+splito :: Term Binary -> Term Binary -> Term Binary -> Term Binary -> Goal ()
+splito n r n1 n2 =
+  disjMany
+    [ do
+        n === inject' []
+        n1 === inject' []
+        n2 === inject' []
+    , do
+        (b, n') <- cons n
+        b === inject' O
+        poso n'
+
+        r === inject' []
+        n1 === n'
+        n2 === inject' []
+    , do
+        (b, n') <- cons n
+        b === inject' I
+
+        r === inject' []
+        n1 === n'
+        n2 === inject' [I]
+    , do
+        (b, n') <- cons n
+        b === inject' O
+        poso n'
+
+        (_, r') <- cons r
+        n2 === inject' []
+        splito n' r' n1 n2
+    , do
+        (b, n') <- cons n
+        b === inject' I
+
+        (_, r') <- cons r
+        n2 === inject' [I]
+        splito n' r' n1 (inject' [])
+    , do
+        (b, n') <- cons n
+        (_, r') <- cons r
+
+        n2' <- fresh
+        poso n2'
+        n2 === Value (LogicCons b n2')
+        splito n' r' n1 n2'
+    ]
+
+trimap :: (a -> b) -> (a, a, a) -> (b, b, b)
+trimap f (x, y, z) = (f x, f y, f z)
+
+quadmap :: (a -> b) -> (a, a, a, a) -> (b, b, b, b)
```

Comment:
Why not
```
data Quad a = Quad a a a a deriving (Functor)
```
?

Reply:
Or `data V4 a = V4 a a a a` (4-dimensional vector).

Reply:
`trimap` and `quadmap ` is only used in these examples:

```hs
putStrLn "\naddo x y z:"
mapM_ print $ take 10 $ trimap tryExtract' <$> run (\(x, y, z) -> addo x y z)
putStrLn "\ndivo n m q r:"
mapM_ print $ take 10 $ quadmap tryExtract' <$> run (\(n, m, q, r) -> divo n m q r)
```

note that 3- and 4-tuples arise because we want to query several variables, and we try to extract the solutions because the output is incomprehensible otherwise. We could go with `V3` and `V4` , but it seems strange to define them and implement `Fresh` for them in this example (especially since we want to move them to doctests). Defining them in the library doesn't seem suitable as well. It may be worth to extend the `Fresh` class with a method like `extractEach`, but at least this is out of scope for this PR.

Reply:
I removed these examples after moving to doctests, so this thread may not be as relevant now.